draft-schinazi-quic-version-negotiation.md

```diff
@@ -112,42 +115,143 @@ versions as long as they are reasonably popular.  This is not generally an issue
+  connection, this field SHALL be 0.
+
+Compatible Version Count:
+
+: A variable-length integer specifying the number of Compatible Version fields
+  following it. The client lists all versions compatible with Attempted Version
+  in the subsequent Compatible Version fields, ordered by descending preference.
+
+~~~
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                    Negotiated Version (32)                    |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                 Supported Version Count (i)                 ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                    Supported Version 1 (32)                   |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                   [Supported Version 2 (32)]                  |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+                               ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                   [Supported Version N (32)]                  |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
+{: #fig-server-tp title="Server Transport Parameter Format"}
+
+The content of each field is described below:
+
+Negotiated Version:
+
+: The version that the server chose to use for the connection. This field SHALL
+  be equal to the value of the Version field in the long header that carries
+  this transport parameter.
+
+Supported Version Count:
+
+: A variable-length integer specifying the number of Supported Version fields
+  following it. The server encodes all versions it supports in the subsequent
+  list, ordered by descending preference.
+
+
+Clients MAY include versions following the pattern `0x?a?a?a?a` in their
+Compatible Version list, and the server in their Supported Version list.
+Those versions are reserved to exercise version negotiation (see the Versions
+section of {{QUIC}}), and MUST be ignored when parsing these fields. Clients
+MUST NOT add reserved versions that it did not oberve in the Version
+Negotiation packet to the Received Negotiation Version list, and clients MUST
+include any reserved versions seen in the Version Negotiation packet to the
+Received Negotiation Version list.
+
+
+# Version Downgrade Prevention
+
+When a server parses the client's `version_negotiation` transport parameter, if
+the `Received Negotiation Version Count` is not zero, the server MUST validate
+that it could have sent the Version Negotation packet described by the client.
+In particular, the server MUST ensure that there are no versions that it
+supports that are absent from the Received Negotiation Versions. This mitigates
+an attacker's ability to forge Version Negotiation packets to force a version
+downgrade.
```

Comment:
Is anything else required.

Reply:
For a version downgrade to be successful, the attacker needs to forge one of
- client's first attempted version
- server's VN packet
- client's second attempted version
- server's negotiated version

This puts all of these in the TLS transcript so as long as they are validated we should be fine?